hazelcast/src/main/java/com/hazelcast/client/impl/ClientPartitionListenerService.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.impl;
+
+import com.hazelcast.client.ClientEndpoint;
+import com.hazelcast.client.impl.protocol.ClientMessage;
+import com.hazelcast.client.impl.protocol.codec.ClientAddPartitionListenerCodec;
+import com.hazelcast.nio.Address;
+import com.hazelcast.nio.Connection;
+import com.hazelcast.spi.impl.NodeEngineImpl;
+import com.hazelcast.spi.partition.IPartition;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class ClientPartitionListenerService {
+
+    private final Map<ClientEndpoint, Long> partitionListeningEndpoints = new ConcurrentHashMap<ClientEndpoint, Long>();
+    private final NodeEngineImpl nodeEngine;
+
+    ClientPartitionListenerService(NodeEngineImpl nodeEngine) {
+        this.nodeEngine = nodeEngine;
+    }
+
+    public void onPartitionStateChange() {
+
+        for (Map.Entry<ClientEndpoint, Long> entry : partitionListeningEndpoints.entrySet()) {
+            ClientMessage clientMessage = getPartitionsMessage();
```

Comment:
can this line be put outside the loop.

Reply:
Since we write to connection it does not directly write to socket but its queued, client message will be shared, and client messages needs to have different correlation id . that is why I need to create new one on each iteration.